_posts/2024-08-19-quarkus-metaprogramming.adoc

```diff
@@ -0,0 +1,372 @@
+
+Finally, also take into account that during the iteration of the fields of a class, the serializers generator also will very likely encounter other types belonging to the application’s business domain. In our original example the `Customer` has an `Address`, a `List<Person>` who are her children and a `CreditCard[]`. When this happens these new types are added to a queue of `ClassInfo` s for which another serializer has to be generated. The whole bytecode generation process terminates only when this queue is empty.
+
+At the end of this process all the newly created classes are passed back to the BuildStep that will record them in the `ResteasyReactiveServerJacksonRecorder` thus making them available also at runtime.
+
+[source, java]
+----
+@BuildStep
+@Record(ExecutionTime.STATIC_INIT)
+public void handleEndpointParams(CombinedIndexBuildItem index,
+ResteasyReactiveServerJacksonRecorder recorder,
+BuildProducer<GeneratedClassBuildItem> generatedClassBuildItemBuildProducer) {
+
+	Map<String, ClassInfo> jsonClasses = new HashMap<>();
+
+	// ... classes to be serialized discovery [omitted]
+
+	if (!jsonClasses.isEmpty()) {
+    	JacksonSerializerFactory factory = new JacksonSerializerFactory(
+            	generatedClassBuildItemBuildProducer, index.getComputingIndex());
+    	factory.create(jsonClasses.values())
+            	.forEach(recorder::recordGeneratedSerializer);
+	}
+}
+----
+
+For this reason it is necessary to add the `@Record` annotation to the BuildStep in order to indicate that it also outputs https://quarkus.io/guides/writing-extensions#bytecode-recording[recorded bytecode]. This concludes the description of all the steps necessary to the implementation of this new feature, that can be visually summarized as it follows.
+
+.Blocks diagram describing the implementation of the generated reflection-free Jackson serializers
+image::DeploymentVsRuntime.jpg[]
+
+== Making it optional
+
+Since it was not possible to reuse any of the heuristics used by Jackson to decide if and how a specific field should be serialized, and as discussed there could still exist some edge cases when the generated serializers produce a result that differs from what Jackson would do using reflection, we decided for now to disable this feature by default and give the possibility to users to opt-in. Considering that all the code implementing this new feature is self-contained in a single BuildStep, that is easily achievable using a https://quarkus.io/guides/writing-extensions#conditional-step-inclusion[conditional step inclusion].
```

Comment:
```suggestion
Since it was not possible to reuse any of the heuristics used by Jackson to decide if and how a specific field should be serialized, and as discussed there could still exist some edge cases when the generated serializers produce a result that differs from what Jackson would do using reflection, we decided for now to disable this feature by default and give the possibility to users to opt-in. Considering that all the code implementing this new feature is self-contained in a single `BuildStep`, that is easily achievable using a https://quarkus.io/guides/writing-extensions#conditional-step-inclusion[conditional step inclusion].
```